lib/puppet/util/character_encoding.rb

```diff
@@ -3,79 +3,72 @@
-          return string.force_encoding(Encoding::UTF_8)
-        elsif currently_valid
-          # If the string is not currently valid UTF-8 but it can be transcoded
-          # (it is valid in its current encoding), we can guess this string was
-          # not originally unicode. Transcode it to UTF-8. For strings with
-          # original encodings like SHIFT_JIS, this should be the final result.
-          return string.encode!(Encoding::UTF_8)
         else
-          # If the string is neither valid UTF-8 as-is nor valid in its current
-          # encoding, fail. It requires user remediation.
-          raise EncodingError
+          # If the string comes to us as BINARY encoded, we don't know what it
+          # started as. However, to encode! we need a starting place, and our
+          # best guess is whatever the system currently is (default_external).
+          # So set external_encoding to default_external before we try to
+          # transcode to UTF-8.
+          string.force_encoding(Encoding.default_external) if original_encoding == Encoding::BINARY
+          string.encode!(Encoding::UTF_8)
         end
       rescue EncodingError => detail
+        # Ensure the string retains it original external encoding since
+        # we've failed
+        string.force_encoding(original_encoding) if original_encoding == Encoding::BINARY
         # Catch both our own self-determined failure to transcode as well as any
         # error on ruby's part, ie Encoding::UndefinedConversionError on a
         # failure to encode!.
-        Puppet.debug(_("%{error}: %{value} is not valid UTF-8 and cannot be transcoded by Puppet.") %
+        Puppet.debug(_("%{error}: %{value} cannot be transcoded by Puppet.") %
           { error: detail.inspect, value: string.dump })
-        return nil
       end
+      return nil
+    end
+
+    # Warning! This is a destructive method - the string supplied is modified!
+    # Given a string, tests if that string's bytes represent valid UTF-8, and if
+    # so sets the external encoding of the string to UTF-8. Does not modify the
+    # byte representation of the string. If the string does not represent valid
+    # UTF-8, does not set the external encoding.
+    #
+    # This method is intended for situations where we do not believe that the
+    # encoding associated with a string is an accurate reflection of its actual
+    # bytes, i.e., effectively when we believe Ruby is incorrect in its assertion
+    # of the encoding of the string.
+    #
+    # @api public
+    # @param [String] string set external encoding (re-label) to utf-8
+    # @return [nil] (string is modified in place)
+    def override_encoding_to_utf_8!(string)
+      if valid_utf_8_bytes?(string)
+        string.force_encoding(Encoding::UTF_8)
+      else
+        Puppet.debug(_("%{value} is not valid UTF-8 and result of overriding encoding would be invalid.") % { value: string.dump })
+      end
+      nil
     end
 
     private
```

Comment:
Can the `valid_bytes` implementation be simplified to?

```ruby
def valid_utf_8_bytes?(string)
  string.dup.force_encoding(Encoding::UTF_8).valid_encoding?
end
```

Given one of your prior PRs, we determined `.dup` doesn't copy a string, but just creates a new wrapper around the same underlying string data - making `dup` pretty cheap.  This lets us test bytes without having to do the save / revert of encoding on the original string.